sw/host/opentitantool/src/command/status_cmd.rs

```diff
@@ -0,0 +1,136 @@
+struct ConsolidateRecord {
+    module_id: String,
+}
+
+impl CommandDispatch for ListCommand {
+    fn run(
+        &self,
+        _context: &dyn Any,
+        _transport: &TransportWrapper,
+    ) -> Result<Option<Box<dyn Annotate>>> {
+        let records = load_elf(&self.elf_file)?;
+        if self.raw_records {
+            return Ok(Some(Box::new(records)));
+        }
+        // Gather things together and do some pretty-printing of module IDs.
+        let filemap = records
+            .into_iter()
+            .try_fold(HashMap::new(), |mut tbl, record| {
+                tbl.entry(record.filename.clone())
+                    .or_insert_with(Vec::new)
+                    .push(ConsolidateRecord {
+                        module_id: record.get_module_id()?,
+                    });
+                Ok::<_, anyhow::Error>(tbl)
+            })?;
+        Ok(Some(Box::new(filemap)))
+    }
+}
+
+#[derive(Debug, StructOpt)]
+/// Lint records in an ELF file.
+pub struct LintCommand {
+    elf_files: Vec<PathBuf>,
+}
+
+#[derive(PartialEq, Eq, Hash, Debug, serde::Serialize)]
+struct ModuleIdProvenance {
+    filename: String,
+    overriden: bool,
+}
+
+impl CommandDispatch for LintCommand {
+    fn run(
+        &self,
+        _context: &dyn Any,
+        _transport: &TransportWrapper,
+    ) -> Result<Option<Box<dyn Annotate>>> {
+        // We group filenames by Module ID, coming from all ELF files at once
+        let mut mod_id_map: HashMap<String, HashSet<ModuleIdProvenance>> = HashMap::new();
+
+        for elf in &self.elf_files {
+            let records = load_elf(elf)?;
+            for record in records {
+                mod_id_map
+                    .entry(record.get_module_id()?)
+                    .or_insert_with(HashSet::new)
+                    .insert(ModuleIdProvenance {
+                        filename: record.filename,
+                        overriden: record.module_id.is_some(),
+                    });
+            }
+        }
+        for (mod_id, prov_set) in mod_id_map.iter() {
+            if prov_set.len() > 1 {
+                println!("The following files have the same module ID ({})", mod_id);
+                for prov in prov_set {
+                    let explain = match prov.overriden {
+                        true => "specified by MAKE_MODULE_ID",
+                        false => "computed from filename",
+                    };
+                    println!("- {}: {}", prov.filename, explain);
+                }
+            }
+        }
+        Ok(None)
+    }
+}
+
+#[derive(Debug, StructOpt)]
+/// Decode a raw status. Optionally accepts an ELF file to recover the filename.
+pub struct DecodeCommand {
+    #[structopt(help = "Raw status to decode (can be in hexadecimal using 0x).", parse(try_from_str = ParseInt::from_str))]
+    raw_status: u32,
+    #[structopt(name = "ELF_FILE", help = "Filename for the executable to analyze.")]
+    _elf_file: Option<PathBuf>,
```

Comment:
Is this just for ability to look at bazel log for filename to inspect?

Reply:
I pushed the wrong version this morning (all the restructuring lead me to loose some code due to an unfortunate git manipulation error). The `decode` command is just to showcase the ability of the library to decode a raw status and find out where a module ID commands from. The plan is to use this in the future when a test hardness communicates with a device test and gets some ujson back with some status error. It is unlikely that the `decode` command itself be used much but it useful to check that the code works.